bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java

```diff
@@ -421,94 +436,187 @@ public LayoutManager getLayoutManager(){
     }
 
     @Override
-    public boolean prepareFormat(ServerConfiguration conf) throws Exception {
-        try (ZooKeeper zk = ZooKeeperClient.newBuilder()
-            .connectString(conf.getZkServers())
-            .sessionTimeoutMs(conf.getZkTimeout())
-            .build()) {
-            boolean ledgerRootExists = null != zk.exists(
-                conf.getZkLedgersRootPath(), false);
-            boolean availableNodeExists = null != zk.exists(
-                conf.getZkAvailableBookiesPath(), false);
-            List<ACL> zkAcls = ZkUtils.getACLs(conf);
-            // Create ledgers root node if not exists
-            if (!ledgerRootExists) {
-                zk.create(conf.getZkLedgersRootPath(), "".getBytes(Charsets.UTF_8),
-                    zkAcls, CreateMode.PERSISTENT);
-            }
-            // create available bookies node if not exists
-            if (!availableNodeExists) {
-                zk.create(conf.getZkAvailableBookiesPath(), "".getBytes(Charsets.UTF_8),
-                    zkAcls, CreateMode.PERSISTENT);
+    public boolean prepareFormat() throws Exception {
+        boolean ledgerRootExists = null != zk.exists(conf.getZkLedgersRootPath(), false);
+        boolean availableNodeExists = null != zk.exists(conf.getZkAvailableBookiesPath(), false);
+        List<ACL> zkAcls = ZkUtils.getACLs(conf);
+        // Create ledgers root node if not exists
+        if (!ledgerRootExists) {
+            zk.create(conf.getZkLedgersRootPath(), "".getBytes(Charsets.UTF_8), zkAcls, CreateMode.PERSISTENT);
+        }
+        // create available bookies node if not exists
+        if (!availableNodeExists) {
+            zk.create(conf.getZkAvailableBookiesPath(), "".getBytes(Charsets.UTF_8), zkAcls, CreateMode.PERSISTENT);
+        }
+
+        // create readonly bookies node if not exists
+        if (null == zk.exists(conf.getZkAvailableBookiesPath() + "/" + READONLY, false)) {
+            zk.create(conf.getZkAvailableBookiesPath() + "/" + READONLY, new byte[0], zkAcls, CreateMode.PERSISTENT);
+        }
+
+        return ledgerRootExists;
+    }
+
+    @Override
+    public boolean initNewCluster() throws Exception {
```

Comment:
the logic here is problematic than the regular format. 

in regular format, it is crashing at the middle of a formatting op, you can re-run format to get it eventually done. in your implementation here, you can't. If I am crashing in the middle of a `init` op, I can't simply re-run `init`.

You might want to document this behavior on those two newly introduced command, to make sure people how to run the commands correctly.

Reply:
an alternative way to use `zk.txn` to make sure all the path creations are done in a multi op. but this probably is out of this scope, can be done in a separate task.

Reply:
I added enough logging when these operations fail, for the admin users to know why exactly it failed. Anyhow I'll add little more information to the issue description, regarding usage of these commands.

Reply:
yes. we need enough instructions for people to run the command when they fail at the middle of running this command.